src/openms/source/ANALYSIS/OPENSWATH/MRMFeatureFilter.cpp

```diff
@@ -0,0 +1,408 @@
+#include <OpenMS/FORMAT/QcMLFile.h>
+
+#include <OpenMS/DATASTRUCTURES/DefaultParamHandler.h>
+#include <OpenMS/CONCEPT/LogStream.h>
+
+namespace OpenMS
+{
+
+  MRMFeatureFilter::MRMFeatureFilter() :
+    DefaultParamHandler("MRMFeatureFilter")
+  {
+    defaults_.setValue("flag_or_filter", "flag", "Flag or Filter (i.e., remove) Components or transitions that do not pass the QC.", ListUtils::create<String>("advanced"));
+    defaults_.setValidStrings("flag_or_filter", ListUtils::create<String>("flag,filter"));
+    
+    //TODO:  future implementation for QcML reporting
+    defaults_.setValue("report_xic", "false", "Embed an image of the XIC in the QC report.", ListUtils::create<String>("advanced"));
+    defaults_.setValidStrings("report_xic", ListUtils::create<String>("true,false"));
+
+    //TODO:  future implementation for QcML reporting
+    defaults_.setValue("report_tic", "false", "Embed an image of the TIC in the QC report.", ListUtils::create<String>("advanced"));
+    defaults_.setValidStrings("report_tic", ListUtils::create<String>("true,false"));
+
+    // write defaults into Param object param_
+    defaultsToParam_();
+    updateMembers_();
+  }
+
+  MRMFeatureFilter::~MRMFeatureFilter()
+  {
+  }
+
+  void MRMFeatureFilter::updateMembers_()
+  {
+    flag_or_filter_ = (String)param_.getValue("flag_or_filter");
+    report_xic_ = param_.getValue("report_xic").toBool();
+    report_tic_ = param_.getValue("report_tic").toBool();
+  }
+
+  void MRMFeatureFilter::FilterFeatureMap(FeatureMap& features, 
+    const MRMFeatureQC& filter_criteria,
+    const TargetedExperiment & transitions
+  )
+  {     
+    // initialize QC variables
+    FeatureMap features_filtered;
+
+    // bool qc_pass;
+    String concentration_units;// iterate through each component_group/feature     
+
+    for (size_t feature_it = 0; feature_it < features.size(); ++feature_it)
+    {      
+      String component_group_name = (String)features[feature_it].getMetaValue("PeptideRef");
+      std::cout << "component_group_name" << component_group_name << std::endl; //debugging
+
+      std::map<String,int> labels_and_transition_types = countLabelsAndTransitionTypes(features[feature_it], transitions);
+
+      // initialize the new feature and subordinates
+      std::vector<Feature> subordinates_filtered;
+      bool cg_qc_pass = true;
+
+      // iterate through each component/sub-feature
+      for (size_t sub_it = 0; sub_it < features[feature_it].getSubordinates().size(); ++sub_it)
+      {
+        String component_name = (String)features[feature_it].getSubordinates()[sub_it].getMetaValue("native_id"); 
+        std::cout << "component_name" << component_name << std::endl; //debugging
+        bool c_qc_pass = true;
+
+        // iterate through multi-feature/multi-sub-feature QCs/filters
+        // iterate through component_groups
+        for (size_t cg_qc_it = 0; cg_qc_it < filter_criteria.component_group_qcs.size(); ++cg_qc_it)
```

Comment:
It may make senseto replace this with a lookup, e.g. store the criteria in a map of `std::map<std::string, *ComponentGroupQC>` or similar